### PR TITLE
New TheGamesDB API support for the scraper

### PR DESCRIPTION
**Pull request by acrummyidea**
Adding new TheGamesDB JSON API scraper, addressing #504 an #516 .
The scraper has all functionality the old scraper had (search by ID search by name, result list etc).
The new api does not directly return the name of publishers and game developers instead it relies on a separate call to get the full list. To avoid downloading such lists I have included two maps containing the relevant information.
To handle JSON responses I have used RapidJson which is now a dependency for the build (I have updated CMake configura files to this end). RapidJson is a headers only library which simplifies things (no new runtime dependencies)
The changes were significant enough that I have created a new class and removed the old scraper to avoid potential reviewers to have to deal with an unwieldy diff.

### CODE REVIEW

**Comment by 5schatten** [comment]
I struggled a bit to apply this PR as .patch file but once I've forked the repo & merged the PR it works quite fine for me. Tried to scrape some games, got the metadata & artwork so I guess it should work.

**Comment by acrummyidea** [comment]
Thanks for testing it!

**Comment by 5schatten** [comment]
No problem I guess we have to thank you for the patch 👍  

I was wondering if it possible to add more artwork to the gamelist? Here is an example of a gamelist created by [UXS ](https://github.com/Universal-Rom-Tools/Universal-XML-Scraper/releases )https://pastebin.com/NVpQmnwc and in addition to the image it downloads an marquee too.

But I guess TGDB lacks additional media in many cases https://thegamesdb.net/game.php?id=50175 ?

**Comment by cmitu** [comment]
Thank you for submitting this. I've compiled the PR on a PI and did a few tests by 
 - individually some ROMs, it seems to work fine. 
 - batch scraping seems to be slow - first entry is scraped fine, then the next one just stalls. I'll try to do more testing for this.
 - when the game's description contains **( something ) **, the text between the parenthesis is replaced with space - see https://imgur.com/a/CxjJTjb

**Comment by joolswills** [comment]
The code styling / indentation does not match that of the rest of the code base. I believe this will need more work before it can be accepted (Feedback from other devs welcome).

It may be better for us to just include the screenscraper PR for now.

**Comment by acrummyidea** [comment]
I used the default settings on clang format. I can clean it up tonight. I won't deny getting this in would be a big motivator for me.

**Comment by 5schatten** [comment]
@cmitu @acrummyidea 
IMHO it would be great if you could combine your efforts & melt your code = profit for us 👍 

An overhauled built-in scraper with the ability to either scrape TGDB or screescraper would be... wait for it... legendary 🥇

**Comment by acrummyidea** [comment]
It's very easy to do since the current code supports multiple scrapers. The PRs can be basically merged sequentially

**Comment by cmitu** [comment]
@5schatten as @acrummyidea said, you can have both and select between them, ES already supports this.

**Comment by 5schatten** [comment]
Ofc ES supports multiple sources if they are available. But I can just select TGDB atm & as I said earlier it would be great if we could not only scrape some information & an image but also the marquee and maybe the video preview too if it's available.

**Comment by cmitu** [comment]
@acrummyidea I've re-tested a bit the batch scraping (scraping a whole system at once) and it manages to finish, it's just that after getting the first 2-3 entries, the next ones seem to be retrieved slowly (a couple minutes ?). The actual info comes into screen (metadata plus thumbnail), it looks like the image download & saving happens slowly before advancing to the next entry. It might be some throttling on the server side though.
All tests were on a stock RetroPie install, Raspberry Pi 3B.

**Comment by acrummyidea** [comment]
@cmitu I suspect throttling. I have reloaded all my ROMs to test but in interactive mode. Did you do it disabling interactive?
@5schatten I'll take a look at what else I can fill in in the gameslist. The original scraper was only getting a thumbnail and boxart. The API gives a list of boxart and fanart per game. If there's room in the ES data structure I can add them.

**Comment by acrummyidea** [comment]
@joolswills I have changed the code to adopt the project's coding style. It should be good to go now
I have prepared a clang-format config as close as I could get it to the style I have seen in the .cpp files in the project. I prepared a gist with it: https://gist.github.com/acrummyidea/0cc539b72459d9b5186e57781dfd50cf
I have used clang-format 7 (the stock clang-format in raspbian is older).
@cmitu the config file may be useful to format you PR as well :)
Also regarding the performance I have tried scraping a whole platform without user input and although it's not fast it did not hang, taking between 1 and 2 seconds per game. I'm pretty sure that TheGamesDB throttles accesses with the public API key.
Down the line we should/could add parallel requests in the scraper to speed things up.

**Comment by acrummyidea** [comment]
@cmitu the issue with the missing text in parenthesis is a UI font issue, I tried the very same rom and it work for me (does show Japanese characters). I checked the corresponding `gamelist.xml` and it’s also ok.

**Comment by cmitu** [comment]
@acrummyidea Yes, I switched the font and the chars are shown, so it's not a scraper problem.

**Comment by Zer0xFF** [comment]
> I'm pretty sure that TheGamesDB throttles accesses with the public API key.

beside the monthly limit, there aren't any restrictions in place.

> Down the line we should/could add parallel requests in the scraper to speed things up.

if you want to speed things up, batch request games with `/ByGameID` when possible, as `id=1,2,3,4,5` can take multiple ids (note each page will only ever contain 20 games, but then just circulate all the returned pages).

we're currently looking into how best to implement serials and potentially crc as well, if you get batching working now, then you could just adapt it to batch request with arc instead later (though yes we're starting with an empty database for those, so it might be a while before that becomes completely useful)

**Comment by acrummyidea** [comment]
@Zer0xFF thanks for your response! Good to know there’s no throttling :)
The main way this scraper is used is to perform a search by name. The search by I’d is a last resort that can be entered manually by users in the UI. 
I’d be curious to know how TheGamesDB backend performs the search by name. We could use a fuzzier search since I’ve noticed that anything that has the article last does not generate matches (made up example: “Flintstones, the” is way less likely to return a result than “The Flintstones” is)

**Comment by Zer0xFF** [comment]
the site code is all up on [github ](https://github.com/TheGamesDB/TheGamesDBv2/blob/dev/include/TGDB.API.php#L263) but we do nothing special with the search term beside an extra phonetic search as well, i know the old code base accounted for article position or as I'd word it manipulated the data, which I'm against, because where do you draw the line? next I'd need to check for prefixed and appended "[]", "{}", "()" etc etc etc....

however, there is a light at the end of the tunnel, mysql has natural search which seems to work well with all the previous mentioned cases, with 1 small issue, you lose order since "order" is something really hard to define in such case, since the base search term could look like this "[RELEASEGROUPNAME] GTA (1987) {blah}"

as shown believe for example
gta_mysql_natual_no_sort
gta_mysql_natual_sort

with that said, I'm happy enough to enable this as i can see a genuine use case for this, though I dont think thats the case here(atleast not long term), as it does work in the case of “Flintstones, the”, though i think in time you should drop it for crc/serials, as they leave little room for guessing.

**Comment by acrummyidea** [comment]
@joolswills this should be ready to merge now. Please let me know if you have any additional comments and I'll be happy to address them :)

**Comment by Zer0xFF** [comment]
natural search has been enabled if you want to test it, `mode=natural`.

**Comment by acrummyidea** [comment]
@Zer0xFF fantastic! I’ll give it a try :)

**Comment by Nikorasu95** [comment]
When will this be merged?

**Comment by tomaz82** [comment]
> When will this be merged?

When it's done

**Comment by acrummyidea** [comment]
@tomaz82 not sure I understand your comment: you mean it as in “it will happen when it happens”, or as in there still work to be done in the PR?

**Comment by tomaz82** [comment]
> @tomaz82 not sure I understand your comment: you mean it as in “it will happen when it happens”, or as in there still work to be done in the PR?

“it will happen when it happens”

**Comment by Lumbeeslayer** [comment]
Seems like this article explains why a lot of people on Github are so rude! https://www.zdnet.com/article/github-open-source-is-dominated-by-men-who-just-cant-communicate/

**Comment by joolswills** [comment]
Looking at the code briefly (and it was mentioned by another developer also), I don't like the hardcoded list of publishers. Having the platforms hardcoded is needed as we map against the ES platforms, but I don't think data like this should be.

**Comment by joolswills** [comment]
My current plan is to get the screenscraper.fr scraper in, and this one after, once it has been adjusted a bit. Appreciate your efforts though.

**Comment by acrummyidea** [comment]
@joolswills I made that choice because the API does not return human readable publishers and developers, just the IDs. And the list is a 300 kb download. I checked the impact on the executable size and it's negligible.
The alternative is to download it and cache it. If you prefer it that way I can modify the PR.

**Comment by cmitu** [comment]
@acrummyidea Skyscraper uses the same data for scraping, but includes the info externally (as `.json` files). That might be an alternative to include it in the code (as it can be updated externally, without re-compiling).

**Comment by acrummyidea** [comment]
@cmitu it's a good idea. Ultimately it comes down to taste and I need to know what @joolswills will like otherwise it'll just be a time sink. I'm happy to help but I also want it to be an effective use of my time.
@joolswills if you have a preference please let me know so as to avoid a trial and error approach.

**Comment by joolswills** [comment]
I think the ideal solution would be for the code to download and perhaps cache it for a day. But if that's more work we could include some xml formatted file, with some script to update it. I'd be interested to hear other developers thoughts though.

**Comment by pjft** [comment]
Downloading and caching is a good option. Would the alternative option a script to update it, via RetroPie-Setup or something? 

That being said, I wouldn't mind having this PR merged for now - as it's usable already and, if I understand correctly, we can then add another PR to add the database handling.

**Comment by tomaz82** [comment]
I'd prefer a download and cache, and have that implemented before a merge, as to not pollute the git with temporary data that will be removed / replaced anyway.

**Comment by Zer0xFF** [comment]
ideally you'd cache it, if you every come across an unknown id, aka a new publisher/devs, you download and cache it again. since you're already parsing and reading json, this should be of little effort to get this part done.(you could, update the list, but I wouldn't even bother, every time you find an unknown id, just request the entire list and update it again)

> Would the alternative option a script to update it, via RetroPie-Setup or something?

you're just asking for trouble now.

**Comment by pjft** [comment]
> you're just asking for trouble now.

...I'm pretty sure "trouble" is written in the reviews of half of my PRs :D

**Comment by acrummyidea** [comment]
@tomasz82 I'll take a stab at it tonight.

@Zer0xFF as a possible improvement of the GamesDb API I think it would be preferable in the future to have the option of requesting text fields for publishers developers and genres in the response. It's hard to imagine a use case where the caller will not end up resolving those numeric IDs that come back.

**Comment by Zer0xFF** [comment]
the decision to make them them IDs was about minimising server load and also the data we have to send out, I think the workload on the client side to resolve these is very reasonable.

**Comment by joolswills** [comment]
I agree that for an API like this, text fields would be good (as an option). Understand server load argument, but it's a small table join and could be cachable server side. But appreciate having the database running well and usable. Thanks for your efforts. :-)

@acrummyidea thanks!

**Comment by acrummyidea** [comment]
@joolswills @tomaz82 I have removed the hard coded lists. The genre, publishers and developers lists are now downloaded if missing and saved in `/tmp`, which is wiped periodically anyway. I had implemented a multi threaded version then found out that `HttpReq` is not thread safe so I went back to a clumsier single threaded version.
The downside is now that downloading the developer list takes a long time, even if it is only 300kb.

**Comment by 5schatten** [comment]
@acrummyidea 
Would it be possible to store the cache in a subfolder so something like `/tmp/emulationstation` or else? Just to have a somewhat "tidy" `/tmp` folder?

**Comment by pjft** [comment]
@5schatten correct. A /tmp subfolder within the binary folder would work (and while at it, the name can be whatever @acrummyidea chooses - if he wants it to be scraper/ or gdb/ or something, it's perfectly fine).

Thanks for putting this together!

**Comment by acrummyidea** [comment]
@pjft @5schatten sounds reasonable. Do you have any pointers on how to reliably find the EmulationStation folder?

**Comment by pjft** [comment]
@acrummyidea this is one example from back in the day to create the folder:

https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/CollectionSystemManager.cpp#L53

and this is the method that gets the actual path:

https://github.com/RetroPie/EmulationStation/blob/master/es-app/src/CollectionSystemManager.cpp#L1043

Even in main.cpp we're using Utils::FileSystem::getHomePath() to get the home folder, and appending /.emulationstation to it to get the configDir.

I'm not sure if there's a better alternative, but given that this is used in multiple places I'd go for it as well.

**Comment by acrummyidea** [comment]
@joolswills, any feedback on the PR?

**Comment by joolswills** [comment]
My apologies. I missed your last update. Will check / test. Anyone else have time to try this also?

**Comment by joolswills** [comment]
If possible if you can rebase and squash that would be good also. One commit or sensibility grouped.

File: es-app/src/scrapers/Scraper.cpp [hunk 1]
```diff
@@ -10,7 +10,7 @@
 #include <fstream>
 
 const std::map<std::string, generate_scraper_requests_func> scraper_request_funcs {
-//	{ "TheGamesDB", &thegamesdb_generate_scraper_requests },
+    { "TheGamesDB", &thegamesdb_generate_json_scraper_requests },
```
Comment (joolswills) [inline es-app/src/scrapers/Scraper.cpp]: minor thing, but wrong indent.

**Comment by joolswills** [comment]
Had a quick look at the new changes and looks good to me although I've had less involvement in the ES code than others. So the Publishers etc are now downloaded once and saved in resources. They are not updated right ? (I couldn't see anything related to refreshing based on date or anything). That's fine of course, but we could clear out this data for example from retropie-setup when updating (which would ensure people have the latest lists).

Feedback from others welcome. Apart from that, once squashed I'm happy to merge. Thanks for working on this.

**Comment by acrummyidea** [comment]
@joolswills correct the resources are downloaded on first use if the files are not present. I agree it’s a good idea to add a script in Retropie setup to delete the resource files on update. I’ll fix the indent you mentioned, rebase and squash, probably tonight.

**Comment by joolswills** [comment]
Thanks!

**Comment by acrummyidea** [comment]
@joolswills done. Thanks for your patience :)

**Comment by joolswills** [comment]
Thanks!

**Comment by Lumbeechief193** [comment]
Hello, I'm having a serious problem with this new merge. I updated my Retropie to the latest version because the issue on this page was merged and said to be fixed. My problem now is all the scrapes I did for my games before TheGameDB API was changed are now completely gone, that means no images, no description, etc. Even the RetroPie section in EmulationStation is missing images and their description now. The files are still in my Retropie directory /opt/retropie/configs/all/emulationstation for my platforms where they should be, they haven't been deleted. I tried scraping with the newest TheGameDB and Skyscraper shown in the EmulationStation but it's basically saying everything has already been scrapped even though the images, and descriptions isn't showing up. I spent a lot of work before the GameDB API was changed, some images that were scrapped were low quality so I downloaded higher quality images and added them to the "downloaded_images" for said platform, now all the hard work I did is in vain.

**Comment by joolswills** [comment]
@Lumbeechief193 please use the retropie forum for support.

**Pull request merged**